### PR TITLE
batched-bench : fix llama_synchronize usage during prompt processing

### PR DESCRIPTION
**Pull request by ggerganov**
During multi-batch prompt processing we were incorrectly synchronizing the llama context on each batch. Instead we just need to synchronize at the end when the prompt is fully computed.

### CODE REVIEW

**Comment by ggerganov** [comment]
@slaren Somewhat related, the logic in `llama-bench` for processing the prompt:

https://github.com/ggml-org/llama.cpp/blob/186415d59552bd5c90549cd8e9be3cc287621fd9/tools/llama-bench/llama-bench.cpp#L1746-L1758

Using `llama_batch_get_one()` here will cause to queue a copy of the logits of the last token in every batch. So for example `-p 8192 -b 2048 -ub 512` will cause 3 extra logit copies at 2048, 4096 and ~6120~6144 processed tokens, compared to only reading the logits for the 8192th token. Setting `-b 8192` or other number larger than `-p` would avoid that, but I am not sure the intent here was to queue these extra copies for the prompt processing test.

**Comment by slaren** [comment]
I was aware of the issue with llama-bench requesting the logits of intermediate batches, I just didn't think that it affects performance enough to worry about it. But I can fix it if that's not the case.

**Comment by ggerganov** [comment]
I haven't timed it, but my guess is that it is indeed quite negligible.

I'll submit a patch for this as I am working on the async Metal backend now and this is a bit related.
